# Lenses Lab

#### _Note: Variable names in the code don't match part numbers given in questions, they refer to the order the experiments (1.1, 2.1, 1.2, 2.2)._

Error for all parts was determined by seeing how far the white background could be moved without noticing the blur. We concluded this value to be about a centimeter.

## Measurements

### 1) Write down your measurement of the image distance from Experiment 1(part 1.1), when your object was at a distance of infinity, including the uncertainty.  (2 points)

Note that the image distance, and focal point are identical for this part, since the light comes from infinity. The measurements gathered for part 1 are printed in the cell below.

In [78]:
from lablib import *
import physplot

fp1 = 16.2
fp1_err = 1
display_result("Results for Part 1.1 (Image Distance/Focal Point):", fp1, fp1_err, "cm")

Results for Part 1.1 (Image Distance/Focal Point): $ 16.20 \pm 1.00 \: cm $

### 2) Write down your measurements of the object and image distances from Experiment 2(part 1.2) for one of your cases including the uncertainties.  (3 points)

Python code displays the first measure object and image distance.

In [106]:
o3s = np.array([1.5 * fp1, 2 * fp1, 4 * fp1])
img3 = np.array([64.8, 61.5, 84.5]) - o3s
markdown("#### A Measurement For Part 1.2 (1.5f or 1st Measurement)")
display_result(
    ["Object Distance for 1st Measurement:", "Image Distance for 1st Measurement"],
    [o3s[0], img3[0]],
    [1, 1], # Again, assume 1 for the image error, 1 for object since it was lined up approximately to the start of the bar...
    "cm"
)

#### A Measurement For Part 1.2 (1.5f or 1st Measurement)

Object Distance for 1st Measurement: $ 24.30 \pm 1.00 \: cm $

Image Distance for 1st Measurement $ 40.50 \pm 1.00 \: cm $

### 3) Write down your measurement of the image distance from Experiment 3(part 2.1), when your object was at a distance of infinity, including the uncertainty.  (2 points)

Python code displays result below. It was about $40 \: cm$.

In [101]:
img2 = 40.3
img2_err = 1
display_result(f"Image Distance for Part 2.1:", img2, img2_err, "cm")

# Calculate focal point here...
fp2 = 1 / ((1 / img2) - (1 / fp1))
fp2_err = 1
# display_result("Results for Part 2.1:", fp2, fp2_err, "cm")

Image Distance for Part 2.1: $ 40.30 \pm 1.00 \: cm $

### 4) Write down your measurement of the object distance from the converging lens that you used in Experiment 4(part 2.2), including the uncertainty. (for separation of lenses of 8 cm)  (2 points)

Code below prints the values for the object distance, (image from part 1.2), from the divergent lens. Notice they are all negative since they are on the same side as the image.

In [81]:
o4s = -(img3[1:] - np.array([15, 8]))
markdown("#### Object Distances for Part 2.2")

img4s = np.array([77.8 - 47.4, 97.2 - 72.8])
display_result(
    [f"Object Disance for {i}f:" for i in [2, 4]],
    o4s,
    1,
    "cm"
)

#### Object Distances for Part 2.2

Object Disance for 2f: $ -14.10 \pm 1.00 \: cm $

Object Disance for 4f: $ -11.70 \pm 1.00 \: cm $

Results for Part 2.2: $ -24.39 \pm 1.00 \: cm $

## Results and Analysis

### 5) What was the focal length you calculated in Experiment 2 (part 1.2) for one of your cases?  Show your work.  (3 points)

Code below was used to calculate the focal distance for experiment 2, part 1.2. Below the cell displays the results for all of the cases, and the averaged value for the focal distance is displayed.

In [85]:
fp3s = 1 / (1 / o3s + 1 / img3)
fp3 = np.mean(fp3s)
fp3_err = 1

markdown("#### Individual Results for Part 1.2")
display_result(
    [f"Focal Distance for {f}f Case:" for f in [1.5, 2, 4]],
    fp3s,
    fp3_err,
    "cm"
)

markdown("#### Average Focal Distance for Part 1.2")
display_result("Combined Result for Part 1.2 (Average):", fp3, fp3_err, "cm")

#### Individual Results for Part 1.2

Focal Distance for 1.5f Case: $ 15.19 \pm 1.00 \: cm $

Focal Distance for 2f Case: $ 15.33 \pm 1.00 \: cm $

Focal Distance for 4f Case: $ 15.11 \pm 1.00 \: cm $

#### Average Focal Distance for Part 1.2

Combined Result for Part 1.2 (Average): $ 15.21 \pm 1.00 \: cm $

### 6) What is the focal length you calculated in Experiment 1 (part 1.1) and the average from Experiment 2 (part 1.2) including the uncertainties?  Do these values agree?  Justify your answer.  (3 points)

The focal distance for part 1.1 was $16.20 \pm 1.00 \: cm$, and the average focal distance for part 2 was $15.21 \pm 1.00 \: cm$. These values (barely) agree, as their uncertainty ranges of $-1$ to $+1$ do actually overlap. The code below checks this on the unrounded values, confirming they agree.

In [93]:
markdown("#### Results from Parts 1.1 and 1.2")
display_result("Results for Part 1.1:", fp1, fp1_err, "cm")
display_result("Combined Result for Part 1.2 (Average):", fp3, fp3_err, "cm")
markdown(f"_Do values agree for parts 1.1 and 1.2? {'Yes!' if(values_agree(fp1, fp1_err, fp3, fp3_err)) else 'No!'}_")

#### Results from Parts 1.1 and 1.2

Results for Part 1.1: $ 16.20 \pm 1.00 \: cm $

Combined Result for Part 1.2 (Average): $ 15.21 \pm 1.00 \: cm $

_Do values agree for parts 1.1 and 1.2? Yes!_

### 7) What was the focal length you calculated for the diverging lens for one of your cases in Experiment 4(part 2.2)?  Show your work.  (4 points)

The following python code below was used to calculate the focal length for part 2.2. Results of these calculations for each part and the average are displayed below.

In [98]:
fp4s = 1 / (1 / o4s + 1 / img4s)
fp4 = np.mean(fp4s)
fp4_err = 1

markdown("#### Individual Results for Part 2.2")
display_result(
    [f"Focal Distance for {f}f Case:" for f in [2, 4]],
    fp4s,
    fp4_err,
    "cm"
)

markdown("#### Average Focal Distance for Part 2.2")
display_result("Combined Result for Part 2.2 (Average):", fp4, fp4_err, "cm")

#### Individual Results for Part 2.2

Focal Distance for 2f Case: $ -26.30 \pm 1.00 \: cm $

Focal Distance for 4f Case: $ -22.48 \pm 1.00 \: cm $

#### Average Focal Distance for Part 2.2

Combined Result for Part 2.2 (Average): $ -24.39 \pm 1.00 \: cm $

###  8) What is the focal length you calculated in Experiment 3(part 2.1) and the average from Experiment 4(part 2.2) including the uncertainties?  Do these values agree?  Justify your answer.  (3 points)

The focal length for part 2.1 is $−27.09 \pm 1.00 \: cm$ and part 2.2 is $−24.39 \pm 1.00 \: cm$. These results happen to not agree with each other, as their uncertainty ranges do not overlap (-27.09 and -24.39 are not 1 away from each other). The code below checks this on the unrounded values and reaches the same conclusion. 

In [103]:
markdown("#### Results for Parts 2.1 and 2.2")
display_result("Results for Part 2.1:", fp2, fp2_err, "cm")
display_result("Combined Result for Part 2.2 (Average):", fp4, fp4_err, "cm")
markdown(f"Do values agree for parts 2.1 and 2.2? {'Yes!' if(values_agree(fp2, fp2_err, fp4, fp4_err)) else 'No!'}")

#### Results for Parts 2.1 and 2.2

Results for Part 2.1: $ -27.09 \pm 1.00 \: cm $

Combined Result for Part 2.2 (Average): $ -24.39 \pm 1.00 \: cm $

Do values agree for parts 2.1 and 2.2? No!

## Concepts

### 9) Why did we have to use a two lens system to find the focal length of the diverging lens?  (3 points)

The image constructed by a divergent lens is virtual, and therefore not visible to the observer (The image appears behind the lens, and does not actually converge at that point, it just appears to diverge from that point). Therefore, to generate a virtual image that is actually visible, the divergent lens must be combined with a convergent lens. The image generated by the convergent lens produces a negative object (object on the same side as the final image), which the divergent lens is capable of showing a visible virtual image for that can be observed, measured.